eos/statistics/log-likelihood.cc

```diff
@@ -879,87 +881,106 @@ namespace eos
+                _observables(gsl_vector_alloc(ids.size())),
+                _observables_2(gsl_vector_alloc(ids.size()))
+            {
+                unsigned k = ids.size();
 
-            // cholesky matrix of covariance
-            std::array<std::array<double, n_>, n_> chol;
+                if (k != _mean->size)
+                    throw InternalError("MultivariateGaussianBlock: dimensions of observables and mean are not identical");
 
-            unsigned _number_of_observations;
-
-            MultivariateGaussianBlock(const ObservableCache & cache, const std::array<ObservableCache::Id, n_> & ids,
-                                     const std::array<double, n_> & mean, const std::array<std::array<double, n_>, n_> & covariance,
-                                     const unsigned & number_of_observations) :
-                cache(cache),
-                ids(ids),
-                mean(mean),
-                covariance(covariance),
-                norm(compute_norm()),
-                _number_of_observations(number_of_observations)
-            {
-                unsigned k = mean.size();
+                if (_covariance->size1 != _covariance->size2)
+                    throw InternalError("MultivariateGaussianBlock: covariance matrix is not a square matrix");
 
-                if ( ids.size() != k || covariance.size() != k|| covariance.front().size() != k)
-                    throw InternalError("MultivariateGaussianBlock.ctor: dimensions of observables, mean and covariance not aligned");
+                if (k != _covariance->size1)
+                    throw InternalError("MultivariateGaussianBlock: dimensions of observables and covariance matrix are not identical");
 
                 // cholesky decomposition (informally: the sqrt of the covariance matrix)
-                // the GSL matrix contains both the cholesky and its transpose, see gsl ref, ch. 14.5
-                gsl_matrix * chol = cholesky();
-                invert_covariance(chol);
+                // the GSL matrix contains both the cholesky and its transpose, see GSL reference, ch. 14.5
+                cholesky();
+                invert_covariance();
 
-                // copy only lower and diagonal part
-                // set upper to zero
-                for (unsigned i = 0 ; i < k ; ++i)
+                // keep only the lower and diagonal parts, set upper parts to zero
+                for (unsigned i = 0; i < k ; ++i)
                 {
-                    for (unsigned j = 0 ; j < k ; ++j)
+                    for (unsigned j = i + 1 ; j < k ; ++j)
                     {
-                        this->chol[i][j] = (j>i) ? 0 : gsl_matrix_get(chol, i, j);
+                        gsl_matrix_set(_chol, i, j, 0.0);
                     }
                 }
-                gsl_matrix_free(chol);
             }
 
             virtual ~MultivariateGaussianBlock()
             {
+                gsl_matrix_free(_covariance_inv);
```

Comment:
RAII. I can see that you don't want to start wrapping lots of GSL structs. But why don't you put them into a `unique_ptr` with a custom deleter?

Reply:
I have commit for this somewhere. Going to recycle it.

Reply:
Ok, it's messier than expected. Either I'd have one header that pulls in basically all of GSL, or many separste ones. I will make this a separate issue to discuss the solution.